### PR TITLE
feat: allow plugins to modify the server and method map

### PR DESCRIPTION
**Pull request by jlipps**
These changes are needed in order to support the new plugin architecture for Appium 2.0. I do not believe they are breaking changes since they just add optional arguments, and change the shape of a method which I don't think is used outside this repo.

### CODE REVIEW

File: lib/express/server.js [hunk 1]
```diff
@@ -93,7 +94,18 @@ async function server (opts = {}) {
...
       });
     });
-    configureServer(app, routeConfiguringFunction, allowCors, basePath);
+
+    configureServer({app, addRoutes: routeConfiguringFunction, allowCors, basePath, plugins});
+
+    // allow plugins to update the app and http server objects
+    for (const plugin of plugins.filter((p) => p.updatesServer)) {
+      log.info(`Allowing plugin ${plugin.name} to modify the Appium server`);
+      plugin.updateServer(app, httpServer).catch(reject);
```
Comment (mykola-mokhnach) [inline lib/express/server.js:103]: should we check first if the plugin has `updateServer` method and this method returns a promise? should this method run asynchronously?

File: lib/protocol/protocol.js [hunk 2]
```diff
@@ -210,13 +210,16 @@ function routeConfiguringFunction (driver) {
     throw new Error('Drivers used with MJSONWP must implement `executeCommand` or `execute`');
   }
 
-  // return a function which will add all the routes to the driver
-  return function addRoutes (app, basePath = DEFAULT_BASE_PATH) {
+  // return a function which will add all the routes to the driver. Here extraMethods might be
+  // passed in as defined by Appium plugins, so we need to add those to the default list
+  return function addRoutes (app, basePath = DEFAULT_BASE_PATH, extraMethods) {
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js:215]: I would rather make everything after `app` argument as options, so the signature would look like `addRoutes(app, ops = {}) `

File: lib/express/server.js [hunk 3]
```diff
@@ -164,16 +182,13 @@ function configureServer (app, routeConfiguringFunction, allowCors = true, baseP
   // set up start logging (which depends on bodyParser doing its thing)
   app.use(startLogFormatter);
 
-  routeConfiguringFunction(app, basePath);
+  addRoutes(app, basePath, plugins.map((p) => p.newMethodMap));
```
Comment (mykola-mokhnach) [inline lib/express/server.js:185]: should we check first if newMethodMap is a plain object?

File: test/express/server-specs.js [hunk 4]
```diff
@@ -14,8 +14,8 @@ describe('server configuration', function () {
   it('should actually use the middleware', function () {
```
Comment (mykola-mokhnach) [inline test/express/server-specs.js:14]: would it be possible to add a unit test for the new feature?

File: lib/express/server.js (hunk 1 above)
Comment (jlipps) [reply lib/express/server.js:103]: Yeah, good idea, for better error messages.

File: lib/protocol/protocol.js (hunk 2 above)
Comment (jlipps) [reply lib/protocol/protocol.js:215]: 👍

File: lib/express/server.js (hunk 3 above)
Comment (jlipps) [reply lib/express/server.js:185]: 👍

File: lib/express/server.js (hunk 1 above)
Comment (jlipps) [reply lib/express/server.js:103]: Oh, and you're right, we want to run those updates synchronously, so I'll need to refactor this a bit.

**Comment by jlipps** [comment]
ok all changes addressed and unit and e2e tests added

File: lib/protocol/protocol.js [hunk 5]
```diff
@@ -210,13 +210,16 @@ function routeConfiguringFunction (driver) {
     throw new Error('Drivers used with MJSONWP must implement `executeCommand` or `execute`');
   }
 
-  // return a function which will add all the routes to the driver
-  return function addRoutes (app, basePath = DEFAULT_BASE_PATH) {
+  // return a function which will add all the routes to the driver. Here extraMethods might be
+  // passed in as defined by Appium plugins, so we need to add those to the default list
+  return function addRoutes ({app, basePath = DEFAULT_BASE_PATH, extraMethods = {}}) {
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js:215]: it makes sense to keep `app` outside of the options dictionary, but up to you to decide

File: test/express/server-specs.js [hunk 6]
```diff
@@ -10,15 +10,76 @@ import sinon from 'sinon';
...
+    const driver = fakeDriver();
+    const addRoutes = routeConfiguringFunction(driver);
+    const plugin = fakePlugin();
+    plugin.newMethodMap = [];
+    configureServer({app: app1, addRoutes});
+    configureServer({app: app2, addRoutes, plugins: [plugin]});
+    app2.totalCount().should.eql(app1.totalCount());
+  });
+
+  it('should allow plugins to update the server', async function () {
```
Comment (mykola-mokhnach) [inline test/express/server-specs.js:80]: should the test be empty?

File: lib/express/server.js [hunk 7]
```diff
@@ -33,13 +33,16 @@ async function server (opts = {}) {
...
   const app = express();
-  let httpServer = http.createServer(app);
-  const serverStateNotifier = new EventEmitter();
-  let isServerClosed = false;
+  const httpServer = http.createServer(app);
+  const serverState = {
+    notifier: new EventEmitter(),
+    closed: false,
+  };
   httpServer.addWebSocketHandler = addWebSocketHandler;
```
Comment (mykola-mokhnach) [inline lib/express/server.js:46]: this all stuff could also be moved to http server configuration

File: lib/express/server.js [hunk 8]
```diff
@@ -50,88 +53,44 @@ async function server (opts = {}) {
...
-                  'other instance of this server running already.');
-      }
+  return await new B(async (resolve, reject) => {
+    // we put an async function as the promise constructor because we want some things to happen in
+    // serial (application of plugin updates, for example). But we still need to use a promise here
+    // because some elements of server start failure only happen in httpServer listeners. So the
+    // way we resolve it is to use an async function here but to wrap all the inner logic in
+    // try/catch so any errors can be passed to reject.
+    try {
+      configureHttp({httpServer, serverState, reject});
```
Comment (mykola-mokhnach) [inline lib/express/server.js:71]: these two synchronous methods could be extracted out of the promise

Comment (jlipps) [reply lib/express/server.js:71]: I don't think so, since they need the `reject` method, right?

Comment (mykola-mokhnach) [reply lib/express/server.js:71]: ah, I see now. Makes sense

File: test/express/server-specs.js (hunk 6 above)
Comment (jlipps) [reply test/express/server-specs.js:80]: whoops, forgot to implement that one.

**Comment by jlipps** [comment]
ok, all comments addressed on this PR.

File: lib/express/server.js [hunk 9]
```diff
@@ -164,16 +101,131 @@ function configureServer (app, routeConfiguringFunction, allowCors = true, baseP
...
+async function configureServerPlugins ({plugins, app, httpServer}) {
+  // allow plugins to update the app and http server objects
+  for (const plugin of plugins.filter((p) => p.updatesServer)) {
+    log.info(`Allowing plugin ${plugin.name} to modify the Appium server`);
+    try {
+      // wrap this in error handling in case the plugin forgot to mark the function async or
+      // forgot to implement it
+      await plugin.updateServer(app, httpServer);
+    } catch (err) {
+      log.error(`Plugin '${plugin.name}' tried to update the server but the updateServer ` +
```
Comment (mykola-mokhnach) [inline lib/express/server.js:191]: it makes sense to write the original error stack into debug log here

Comment (jlipps) [reply lib/express/server.js:191]: stack will be printed when the original error is thrown in the main appium process

**Pull request merged**